mesonbuild/compilers/rust.py

```diff
@@ -191,6 +192,18 @@ def get_pie_args(self) -> T.List[str]:
         # pic is on by rustc
         return []
 
+    def get_lto_compile_args(self, *, threads: int = 0, mode: str = 'default') -> T.List[str]:
+        # Enable cross langauge lto
+        args = ['-C', 'linker-plugin-lto=yes']
```

Comment:
I forgot to mention that `-C linker-plugin-lto=yes` doesn't seem to be compatible with `gcc`: https://github.com/systemd/systemd/pull/19598#discussion_r633071853

In a script compiling systemd on GitHub Actions `-C linker-plugin-lto=yes` is passed only when `clang` is used:
```
if [[ "$COMPILER" == "clang" ]]; then
      ARGS+=("--optimization=3 -Db_lto=true -Drust_args='--deny warnings -Copt-level=3 -Clinker-plugin-lto'")
else
      ARGS+=("--optimization=3 -Db_lto=true -Drust_args='--deny warnings -Copt-level=3'")
fi
```

Reply:
Yeah, I saw that too. I couldn't come up with any incantation that would do any kind of lto with gcc support, and I thought that gcc and clang's lto implementations were not generally compatible anyway? I could be wrong thought that basically gcc acted on gimpl, and llvm acted on LLVM-ir, that they embed in the .o files.

Reply:
I have to say I haven't figured out how `lto`, `linker`, `linker-flavor` and `linker-plugin-lto` interact with each other and why, for example,
```
meson -Drust_args="-C lto=yes"  -Db_lto=true -Dbuild-rust=true build
```
works with both `gcc` and `clang` without `-Clinker-plugin-lto`. Would it be possible to put the "lto" commits on hold (so that it was possible for the systemd CI to keep building systemd with `-Db_lto=true` with both `gcc` and `clang`?

Reply:
Okay, I finally found the actual documentation. without -C linker-plugin-lto rustc performs LTO on the .o file it generates, but then does no interprocdedural optimization at link time (calling it IPO makes more sense than LTO, I guess ICC was smarter than GCC and Clang on this). Defering will result in better optimization, but requires that all of the compielrs be LLVM based and use LTO.

What a pain.

Reply:
I've dropped the LTO commit, I'll try to sort that out later. Now I get the fun of starting up Windows.

Reply:
I can imagine a scenario where I need both "IPO" and the IPO deferred to the link time in C/Rust projects. Now I'm puzzled. I'm not sure what `-Db_lto=true` should do.

Reply:
You can always manually set -D${lang}_args and -D${lang}_link_args and ignore the builtins, but yeah, I think the goal of `-Db_lto` is to provide a mechanism that works as much as possible, and provides the thing most people want. It sounds like from the docs though that IPO at object generation time (or really, when rust takes the .rs files and outputs a .a, .so, etc) is strictly a subset of what happens with link-time IPO.

Reply:
So, I don't think you actually need both, use link-time IPO if possible, but .o time is better than no IPO. Rustc does the .o IPO by default.